Colors/Pallet/Pallet.cs

```diff
@@ -6,79 +6,107 @@
-            else
-            {
-                throw new FileNotFoundException($"File {fileName} does not exist!!!");
-            }
+            } 
         }
 
-        public IEnumerable<KeyValuePair<Color, int>> GetMostFrequentColors(int number)
+        public Pallet(Bitmap bmp)
         {
-            var colors = GetAllColors();
-            return colors.OrderByDescending(c => c.Value).Take(number);
+            _bmp = bmp;
+            //InitColors();
+            InitColorsForBiggerPicturesBruteForce();
         }
 
-        IDictionary<Color, int> GetAllColors()
+        public Pallet(Bitmap bmp, int downgradeSquares)
         {
-            var colors = new Dictionary<Color, int>();
+            _bmp = bmp;
+            InitColorsWithDownGrade(downgradeSquares);
+        }
 
+        public IEnumerable<ColorAbundance> GetMostCommonColors(int number)
+        {
+            return _colors.OrderByDescending(c => c.Abundance).Take(number);
+        }
+
+        void InitColorsForSmallPicturesOnlyBruteForce() // Looks nice, but awfully slow for bigger pictures
+        {
             if (_bmp != null)
             {
                 for (int y = 0; y < _bmp.Size.Height; y++)
                 {
                     for (int x = 0; x < _bmp.Size.Width; x++)
                     {
                         var pixel = _bmp.GetPixel(x, y);
-                        if (colors.ContainsKey(pixel))
-                        {
-                            colors[pixel] = colors[pixel] + 1;
-                        }
-                        else
-                        {
-                            colors[pixel] = 1;
-                        }
+
+                        this[pixel]++;
                     }
                 }
             }
-
-            return colors;
         }
 
-        public int CountImageColors()
+        void InitColorsForBiggerPicturesBruteForce() // Faster because dictionaries work with HashCodes
         {
-            var count = 0;
-            var colors = new HashSet<Color>();
+            var colors = new Dictionary<Color, int>();
 
             if (_bmp != null)
             {
                 for (int y = 0; y < _bmp.Size.Height; y++)
                 {
                     for (int x = 0; x < _bmp.Size.Width; x++)
                     {
-                        colors.Add(_bmp.GetPixel(x, y));
+                        var pixel = _bmp.GetPixel(x, y);
+                        if (colors.ContainsKey(pixel))
+                        {
+                            colors[pixel] = colors[pixel] + 1;
+                        }
+                        else
+                        {
+                            colors[pixel] = 1;
+                        }
                     }
                 }
-                count = colors.Count;
             }
 
-            return count;
+            foreach (var color in colors)
+            {
+                _colors.Add(new ColorAbundance(color.Key, color.Value));
```

Comment:
can we do this in line 96 after colors.Add()? and we can get rid of this extra loop